base_geo_route_optimization/models/geo_route_optimize.py

```diff
@@ -0,0 +1,190 @@
+# Copyright (C) 2020 Brian McMaster
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+import logging
+import json
+import urllib
+
+from ortools.constraint_solver import routing_enums_pb2
+from ortools.constraint_solver import pywrapcp
+
+from odoo import _, api, models
+from odoo.exceptions import ValidationError
+
+_logger = logging.getLogger(__name__)
+
+
+class GeoRouteOptimize(models.AbstractModel):
+    """ This model is meant to solve the Traveling Salesman Problem
+       and/or the Vehicle Routing Problem.
+       It is meant to be inherited by any model that needs to efficiently
+       optimize the sequence in which records are ordered, specifically for
+       driving purposes.
+       Distances between records are calculated and stored in a matrix that is
+       generated by the Google Distance Matrix API
+       The optimization problem is solved using Google OR Tools python package
+       The functions in this model are based on the examples provided in the
+       following link.
+       More information at https://developers.google.com/optimization/routing
+    """
+
+    _name = "geo.route.optimize"
+    _description = "Geo Route Optimization Framework"
+
+    def send_request(self, origin_addresses, dest_addresses):
+        """ Build and send the API request for the given origin and
+           destination addresses.
+        """
+        ICPSudo = self.env["ir.config_parameter"].sudo()
+        api_key = ICPSudo.get_param("google.api_key_distance_matrix")
+        if not api_key:
+            raise ValidationError(_("No Google Distance Matrix API key defined"))
+        request = (
+            "https://maps.googleapis.com/maps/api/distancematrix/json?units=imperial"
+        )
+        origin_address_str = "|".join(origin_addresses)
+        dest_address_str = "|".join(dest_addresses)
+        request = (
+            request
+            + "&origins="
+            + origin_address_str
+            + "&destinations="
+            + dest_address_str
+            + "&key="
+            + api_key
+        )
+        try:
+            jsonResult = urllib.request.urlopen(request).read()
+            response = json.loads(jsonResult)
+        except urllib.error.URLError as e:
+            raise ValidationError(
+                _("Failure when making Distance Matrix API request: %s", e)
+            )
+        return response
+
+    def build_distance_matrix(self, response):
+        """The following function builds rows of the distance matrix,
+           using the response returned by the send_request function."""
+        distance_matrix = []
```

Comment:
I think you must check the status of the response before building the matrix
may be somthing like :
`        if response["status"] != 'OK':
            raise ValidationError(_("Service Api responded with : %s") % response["status"])`